packages/@lwc/ssr-compiler/src/compile-template/transformers/lwc-component.ts

```diff
@@ -0,0 +1,103 @@
+        yield* ${/*component ctor*/ is.expression}[SYMBOL__GENERATE_MARKUP](null, childProps, childAttrs);
+    }
+`<EsBlockStatement>;
+
+const bThrowErrorForInvalidConstructor = esTemplate`
+    {
+        throw new Error(\`Invalid constructor \${String(${/*component ctor*/ is.expression})} is not a LightningElement constructor.\`)
+    }
+`<EsBlockStatement>;
+
+function bIfLwcIsExpressionDefined(lwcIsExpression: Expression, consequent: Statement) {
+    // instance.lwcIsValue !== undefined && instance.lwcIsValue !== null
+    const lwcIsExpressionDefined = b.logicalExpression(
+        '&&',
+        b.binaryExpression('!==', lwcIsExpression, b.identifier('undefined')),
+        b.binaryExpression('!==', lwcIsExpression, b.identifier('null'))
+    );
+
+    return b.ifStatement(lwcIsExpressionDefined, b.blockStatement([consequent]));
+}
+
+function bifLwcIsExpressionTypeCorrect(
+    lwcIsExpression: Expression,
+    consequent: Statement,
+    alternate: Statement
+) {
+    // typeof instance.lwcIsValue === 'function'
+    const typeComparison = b.binaryExpression(
+        '===',
+        b.unaryExpression('typeof', lwcIsExpression),
+        b.literal('function')
+    );
+
+    // instance.lwcIsValue.prototype instanceof LightningElement
+    const protoComparison = b.binaryExpression(
+        'instanceof',
+        b.memberExpression(lwcIsExpression, b.identifier('prototype')),
+        b.identifier('LightningElement')
+    );
+
+    const comparison = b.logicalExpression('&&', typeComparison, protoComparison);
+
+    return b.ifStatement(comparison, consequent, alternate);
+}
+
+export const LwcComponent: Transformer<IrLwcComponent> = function LwcComponent(node, cxt) {
+    const { directives } = node;
+
+    const lwcIs = directives.find((directive) => directive.name === 'Is');
+    if (!isUndefined(lwcIs)) {
+        cxt.hoist(bImportDeclaration(['LightningElement']), 'import:LightningElement');
+        cxt.hoist(
+            bImportDeclaration(['SYMBOL__GENERATE_MARKUP']),
+            'import:SYMBOL__GENERATE_MARKUP'
+        );
+        cxt.hoist(
+            bImportDeclaration([{ cloneAndDeepFreeze: '__cloneAndDeepFreeze' }]),
+            'import:cloneAndDeepFreeze'
+        );
+
+        // The template compiler has validation to prevent lwcIs.value from being a literal
+        const lwcIsExpression = expressionIrToEs(lwcIs.value as IrExpression, cxt);
+        return [
+            bIfLwcIsExpressionDefined(
+                lwcIsExpression,
+                bifLwcIsExpressionTypeCorrect(
+                    lwcIsExpression,
+                    bYieldFromDynamicComponentConstructorGenerator(
+                        getChildAttrsOrProps(node.properties, cxt),
+                        getChildAttrsOrProps(node.attributes, cxt),
+                        lwcIsExpression
+                    ),
+                    bThrowErrorForInvalidConstructor(lwcIsExpression)
+                )
+            ),
+        ];
```

Comment:
Here's a sample output:

```js
if (instance.customCtor !== undefined && instance.customCtor !== null) {
    if (typeof instance.customCtor === "function" && instance.customCtor.prototype instanceof LightningElement) {
      const childProps = cloneAndDeepFreeze({});
      const childAttrs = {};
      yield* instance.customCtor[SYMBOL__GENERATE_MARKUP](null, childProps, childAttrs);
    } else {
      throw new Error(`Invalid constructor ${String(instance.customCtor)} is not a LightningElement constructor.`);
    }
  }
```

Reply:
nit: It's not immediately apparent to me why a bunch little builders were used instead of a larger `esTemplate`. This seems to work fine though - just not quite as readable.

Reply:
I wasn't sure what parts we wanted to keep in an `esTemplate` vs using a builder but this helps clear things up!

I'll redo this to use an `esTemplate`, I think it'll be easier to read as well!

Reply:
[Updated](https://github.com/salesforce/lwc/pull/4847/files#diff-4f3f68417e36c5fccd5449c06afb5813bc1e5000dbc0950302528a8a7536e4dcR23-R36)!